File: packages/typescript/shared/src/utils/encoder-decoder.ts
import { VNF, VNFUpdated } from "../models/embedding-graph";

/**
 * This function encodes the Embedding Graph to a base64 string.
 *
 * @param data The Embedding Graph to be encoded.
 * @returns base64 encoded string
 */
export function sfcEncode(data: VNF | VNF[] | VNFUpdated[]): string {

    const jsonData: string = JSON.stringify(data);
    const base64Data: string = btoa(jsonData)

    return base64Data;
}

/**
 * This function decodes the base64 encoded string to an Embedding Graph.
 *
 * @param data The base64 encoded string to be decoded.
 * @returns The decoded Embedding Graph.
 */
export function sfcDecode(data: string): VNF | VNF[] | VNFUpdated[] {
    const jsonString: string = atob(data);
    const eg: VNF = JSON.parse(jsonString);

    return eg;
}

File: packages/python/shared/utils/encoder_decoder.py
"""
Provides the encoding and decoding functions to encode/decode the SFC metadata in the HTTP headers.
"""

import json
import base64
from typing import Any, Union
from shared.models.embedding_graph import VNF


def sfcEncode(data: "Union[VNF, list[VNF]]") -> Any:
    """
    Encode the SFC metadata in the HTTP headers.

    Parameters:
        data (Union[VNF, list[VNF]]): SFC metadata to be encoded

    Returns:
        Any: Base64-encoded data
    """

    jsonData: str = json.dumps(data)
    base64Data: Any = base64.b64encode(jsonData.encode("ascii")).decode("ascii")

    return base64Data


def sfcDecode(data: Any) -> "Union[VNF, list[VNF]]":
    """
    Decode the SFC metadata in the HTTP headers.

    Parameters:
        data (Any): Base64-encoded data

    Returns:
        dict(Union[VNF, list[VNF]]): Decoded SFC metadata
    """

    jsonData: str = base64.b64decode(data.encode("ascii")).decode("ascii")
    dataDict: VNF = json.loads(jsonData)

    return dataDict
